tt_eager/tt_dnn/op_library/moreh_nll_loss_backward/moreh_nll_loss_backward/kernels/reader_moreh_nll_loss_backward_2d.cpp

```diff
@@ -0,0 +1,146 @@
+
+    const uint32_t divisor_tile_bytes = get_tile_size(cb_divisor);
+    const DataFormat divisor_data_format = get_dataformat(cb_divisor);
+
+    const uint32_t output_grad_tile_bytes = get_tile_size(cb_output_grad);
+    const DataFormat output_grad_data_format = get_dataformat(cb_output_grad);
+
+    const uint32_t target_tile_bytes = get_tile_size(cb_target);
+
+    constexpr bool target_is_dram = get_compile_time_arg_val(0) == 1;
+    constexpr bool weight_is_dram = get_compile_time_arg_val(1) == 1;
+    constexpr bool divisor_is_dram = get_compile_time_arg_val(2) == 1;
+    constexpr bool output_grad_is_dram = get_compile_time_arg_val(3) == 1;
+
+    const InterleavedAddrGen<target_is_dram> addrg_target = {
+        .bank_base_address = target_addr, .page_size = target_tile_bytes};
+    const InterleavedAddrGenFast<output_grad_is_dram> addrg_output_grad = {
+        .bank_base_address = output_grad_addr,
+        .page_size = output_grad_tile_bytes,
+        .data_format = output_grad_data_format};
+    constexpr uint32_t onetile = 1;
+
+#if defined(WEIGHT)
+    const InterleavedAddrGen<weight_is_dram> addrg_weight = {
+        .bank_base_address = weight_addr,
+        .page_size = 1024 * element_size,
+    };
+
+    cb_reserve_back(cb_weight, weight_num_tile);
+    uint32_t l1_write_addr_weight = get_write_ptr(cb_weight);
+    volatile tt_l1_ptr uint16_t* weight_l1_ptr = reinterpret_cast<volatile tt_l1_ptr uint16_t*>(l1_write_addr_weight);
+
+    for (uint32_t i = 0; i < weight_num_tile * 2; ++i) {
+        uint32_t noc_id = i / 2;
+        uint32_t noc_offset = 0;
+        if (noc_id * 2 != i) {
+            noc_offset += 256 * element_size;
+        }
+        uint64_t src_noc_addr = get_noc_addr(noc_id, addrg_weight, noc_offset);
+        noc_async_read(src_noc_addr, l1_write_addr_weight, NOC_MINIMUM_READ_SIZE);
+        noc_async_read_barrier();
+        l1_write_addr_weight += NOC_MINIMUM_READ_SIZE;
+    }
+#endif
+
+#if defined(DIVISOR)
+    const InterleavedAddrGenFast<divisor_is_dram> addrg_divisor = {
+        .bank_base_address = divisor_addr, .page_size = divisor_tile_bytes, .data_format = divisor_data_format};
+
+    cb_reserve_back(cb_divisor, onetile);
+    uint32_t l1_write_addr_divisor = get_write_ptr(cb_divisor);
+    noc_async_read_tile(0, addrg_divisor, l1_write_addr_divisor);
+    noc_async_read_barrier();
+    cb_push_back(cb_divisor, onetile);
+#endif
+
+    cb_reserve_back(cb_output_grad, onetile);
+    uint32_t l1_write_addr_output_grad = get_write_ptr(cb_output_grad);
+    noc_async_read_tile(0, addrg_output_grad, l1_write_addr_output_grad);
+    noc_async_read_barrier();
+    cb_push_back(cb_output_grad, onetile);
+
+
+    uint32_t Ct = (C + TILE_HEIGHT - 1) / TILE_HEIGHT;
+
+    uint32_t end_id = start_id + num_tiles_per_core;
+    for (uint32_t i = start_id; i < end_id; ++i) {
+        uint32_t nt = i / Ct;
+        int32_t ct = static_cast<int32_t>(i % Ct);
```

Comment:
Why int and not uint?

Reply:
It was a mistake to use int32_t. It has been corrected to uint32_t.

Reply:
There are still other instances of this I think. Can you clean those up as well?

Reply:
Yes,  I've cleaned up similar cases in this commit.
```C++
        int32_t c = static_cast<int32_t>(nc % C);
->
        uint32_t c = nc % C;
```


https://github.com/tenstorrent/tt-metal/pull/8966/commits/373a9fed99806c2bc4be03a588ba987530a26d42#diff-2a1285f3fc0f375940074cef369b396765945f23f0070a3865001b0d758d6f87L99